pyro/distributions/lkj.py

```diff
@@ -0,0 +1,205 @@
+
+        z = torch.cat(z_stack, -1)
+        return torch.log1p((2*z)/(1-z)) / 2
+
+    def log_abs_det_jacobian(self, x, y):
+        # Note dependence on pytorch 1.0.1 for batched tril
+        tanpart = x.cosh().log().sum(-1).mul(-2)
+        matpart = (1 - y.pow(2).cumsum(-1).tril(diagonal=-2)).log().div(2).sum(-1).sum(-1)
+        return tanpart + matpart
+
+# register transform to global store
+@biject_to.register(corr_cholesky_constraint)
+@transform_to.register(corr_cholesky_constraint)
+def _transform_to_corr_cholesky(constraint):
+    return CorrLCholeskyTransform()
+
+
+########################################
+# Define distribution
+########################################
+
+# TODO: Modify class to support more than one eta value at a time?
+class CorrLCholeskyLKJPrior(TorchDistribution):
+    """
+    Generates cholesky factors of correlation matrices using an LKJ prior.
+
+    The expected use is to combine it with a vector of variances and pass it
+    to the scale_tril parameter of a multivariate distribution such as MultivariateNormal.
+
+    E.g., if theta is a (positive) vector of covariances with the same dimensionality
+    as this distribution, and Omega is sampled from this distribution,
+    scale_tril=torch.mm(torch.diag(sqrt(theta)), Omega)
+
+    Note that the `event_shape` of this distribution is `[d, d]`
+
+    Important note: When using this distribution with HMC/NUTS, it is important to
+    use a `step_size` such as 1e-4. If not, you are likely to experience LAPACK
+    errors regarding positive-definiteness.
+
+    :param int d: Dimensionality of the matrix
+    :param torch.Tensor eta: A single positive number parameterizing the distribution.
+    """
+    arg_constraints = {"eta": constraints.positive}
+    support = corr_cholesky_constraint
+    has_rsample = True
+
+    def __init__(self, d, eta, validate_args=None):
+        if not torch.is_tensor(eta):
+            raise ValueError("Eta must be a tensor")
+        if any(eta <= 0):
+            raise ValueError("eta must be > 0")
+        if eta.numel() != 1:
+            raise ValueError("eta must be a single number; for a larger batch size, call expand")
+        vector_size = (d * (d - 1)) // 2
+        alpha = eta.add(0.5 * (d - 1.0))
+
+        concentrations = eta.new_empty(vector_size,)
+        i = 0
+        for k in range(d-1):
+            alpha -= .5
+            for j in range(k+1, d):
+                concentrations[..., i] = alpha
+                i += 1
+        self._gen = Beta(concentrations, concentrations)
+        self.eta = eta
+        self._d = d
+        self._lkj_constant = None
+        super(CorrLCholeskyLKJPrior, self).__init__(torch.Size(), torch.Size((d, d)), validate_args=validate_args)
+
+    def rsample(self, sample_shape=torch.Size()):
+        y = self._gen.rsample(sample_shape=self.batch_shape + sample_shape).detach()
+        z = y.to(dtype=self.eta.dtype).mul(2).add(-1.0)
```

Comment:
Is the casting to eta's dtype necessary? I would think that `self._gen` was initialized with eta's `dtype` and would therefore return the correct `dtype`. 

Also, I think this method should be `sample` and `has_rsample = False`, since the distribution isn't reparametrized?

Reply:
Regarding `sample` vs. `rsample`, that's what I thought at first, to use `sample`. But, looking at some of the example distributions, it seems that for, e.g., Beta, the only reparameterization accepted by `rsample` is the `sample_shape`, so doing it this way seemed like the shortest path to complying with pyro's shape semantics.  See, e.g., https://github.com/pytorch/pytorch/blob/016f21235721d9998b1ebcd5c7948e1049233771/torch/distributions/beta.py#L57   

Whichever method you prefer is fine with me, just let me know. 

Regarding the `dtype`, I added that when I was working through the eps issue. I'll check if its still necessary and take it out if its not.  

Thanks!

Reply:
> Beta, the only reparameterization accepted by `rsample` is the `sample_shape`.

The shape signature for `rsample` is the same as `sample` (i.e. it only takes an additional `sample_shape` arg), except with the additional requirement that the samples are differentiable with respect to the distribution parameters. In case of Beta, you should be able to compute the gradient w.r.t. `concentration1` and `concentration0` of any objective that uses a sample generated via `rsample`. With `sample`, you aren't expected to compute gradients through the samples and will likely see an exception when you try to do so.

Reply:
Since we can't do this with the lkj distribution, we should set `has_rsample = False` and just have a non-reparametrized `sample` method.

Reply:
Ok. While it wouldn't be particularly difficult to propagate the gradient through `rsample` (although at a not insignificant cost to performance), I think we should be discouraging users from trying to optimize `eta`.   Thanks!

Reply:
I’m now thinking this about this in detail and I think there’s an interesting question. I hope you won’t mind talking it through with me?

While optimizing eta wouldn’t make sense, what could make sense is optimizing the parameters to the generating set of beta distributions. LKJ by design uses betas where the a and b parameters are equal. If we separate those into two vectors, then we could optimize back. 

The optimized distribution would not be an lkj prior. But it would be a generating distribution for (I think) the posterior of the cholesky factor of the correlation matrix.

What do you think?

If we do this, then how would this need to interact with expand and with sampling more than one correlation matrix at a time?  Would we need to expand and detach the paramaters for the generating distribution so each item in the batch is optimized separately?

Reply:
I think you are right that we might be able to differentiate back maybe with some modifications to the code (I am not completely sure), but I am unclear about the use case you mention. Generally `rsample` is very useful for Variational Inference where the ELBO objective has some expectation term over the latent which we need to differentiate through. We can use the reparametrization trick if our distribution has an `rsample` method, and push the gradient through the expectation to get a low variance gradient estimator. One reason to reparametrize this might be to optimize eta of the variational distribution to maximize the ELBO in VI. I am not sure why we would need to take gradients through a sample otherwise, say in the case of MCMC. Just `log_prob` being differentiable should be sufficient, but maybe I don't fully understand the use case you mention? If we want to attempt to reparametrize LKJ, my suggestion would be to do this in a following PR, just so we can contain the scope of this, and add further tests for `rsample`. 😄

Reply:
Sounds good, thanks!